File: apps/android_native/app/src/main/java/com/example/passwordmanagernative/sync/SyncSettings.kt
package com.example.passwordmanagernative.sync

import org.json.JSONArray
import org.json.JSONObject
import java.time.Instant
import java.util.Locale
import java.util.UUID

enum class SyncProviderType {
    NONE,
    WEBDAV,
    S3_PRESIGNED,
    NAS_WEBDAV,
    TENCENT_COS,
    ALIYUN_OSS;

    val wireName: String
        get() = when (this) {
            NONE -> "none"
            WEBDAV -> "webdav"
            S3_PRESIGNED -> "s3Presigned"
            NAS_WEBDAV -> "nasWebdav"
            TENCENT_COS -> "tencentCos"
            ALIYUN_OSS -> "aliyunOss"
        }

    val title: String
        get() = when (this) {
            NONE -> "None"
            WEBDAV -> "WebDAV"
            S3_PRESIGNED -> "S3 Presigned URL"
            NAS_WEBDAV -> "NAS WebDAV"
            TENCENT_COS -> "Tencent Cloud COS"
            ALIYUN_OSS -> "Alibaba Cloud OSS"
        }

    companion object {
        fun fromWireName(value: String?): SyncProviderType =
            entries.firstOrNull { it.wireName == value } ?: NONE
    }
}

enum class SyncSettingsConflictStrategy {
    REMOTE_WINS,
    LOCAL_WINS,
    KEEP_BOTH;

    val wireName: String
        get() = when (this) {
            REMOTE_WINS -> "remoteWins"
            LOCAL_WINS -> "localWins"
            KEEP_BOTH -> "keepBoth"
        }

    val title: String
        get() = when (this) {
            REMOTE_WINS -> "Remote Wins"
            LOCAL_WINS -> "Local Wins"
            KEEP_BOTH -> "Keep Both"
        }

    val mergeStrategy: SyncConflictStrategy
        get() = when (this) {
            REMOTE_WINS -> SyncConflictStrategy.REMOTE_WINS
            LOCAL_WINS -> SyncConflictStrategy.LOCAL_WINS
            KEEP_BOTH -> SyncConflictStrategy.KEEP_BOTH
        }

    companion object {
        fun fromWireName(value: String?): SyncSettingsConflictStrategy =
            entries.firstOrNull { it.wireName == value } ?: REMOTE_WINS
    }
}

enum class SyncIntervalUnit {
    SECONDS,
    MINUTES;

    val wireName: String
        get() = when (this) {
            SECONDS -> "seconds"
            MINUTES -> "minutes"
        }

    companion object {
        fun fromWireName(value: String?): SyncIntervalUnit =
            entries.firstOrNull { it.wireName == value } ?: MINUTES
    }
}

data class SyncLogEntry(
    val timestamp: Instant,
    val message: String,
    val level: String,
)

data class SyncSettings(
    val providerType: SyncProviderType,
    val webdavUrl: String,
    val webdavUsername: String,
    val webdavPassword: String,
    val webdavPath: String,
    val presignedDownloadUrl: String,
    val presignedUploadUrl: String,
    val objectStorageAccessKeyId: String,
    val objectStorageSecretAccessKey: String,
    val objectStorageBucket: String,
    val objectStorageEndpoint: String,
    val objectStorageAppId: String,
    val objectStorageCustomUrl: String,
    val objectStorageObjectKey: String,
    val autoSyncEnabled: Boolean,
    val autoSyncIntervalMinutes: Int,
    val autoSyncIntervalValue: Int,
    val autoSyncIntervalUnit: SyncIntervalUnit,
    val autoSyncOnUnlock: Boolean,
    val conflictStrategy: SyncSettingsConflictStrategy,
    val syncMasterKey: Boolean,
    val deviceId: String,
    val lastSyncRevision: Int,
    val lastSyncAt: Instant?,
    val lastSyncStatus: String?,
    val lastSyncMessage: String?,
    val lastRemoteFingerprint: String?,
    val hasLocalChanges: Boolean,
    val logs: List<SyncLogEntry>,
) {
    val syncSecrets: SyncSecretBundle
        get() = SyncSecretBundle(
            webdavPassword = webdavPassword,
            presignedDownloadUrl = presignedDownloadUrl,
            presignedUploadUrl = presignedUploadUrl,
            objectStorageAccessKeyId = objectStorageAccessKeyId,
            objectStorageSecretAccessKey = objectStorageSecretAccessKey,
        )

    fun redactedForPlaintextStorage(): SyncSettings =
        applyingSecrets(SyncSecretBundle.EMPTY)

    fun applyingSecrets(secrets: SyncSecretBundle): SyncSettings =
        copy(
            webdavPassword = secrets.webdavPassword,
            presignedDownloadUrl = secrets.presignedDownloadUrl,
            presignedUploadUrl = secrets.presignedUploadUrl,
            objectStorageAccessKeyId = secrets.objectStorageAccessKeyId,
            objectStorageSecretAccessKey = secrets.objectStorageSecretAccessKey,
        )

    fun toJson(): JSONObject =
        JSONObject()
            .put("providerType", providerType.wireName)
            .put("webdavUrl", webdavUrl)
            .put("webdavUsername", webdavUsername)
            .put("webdavPassword", webdavPassword)
            .put("webdavPath", webdavPath)
            .put("presignedDownloadUrl", presignedDownloadUrl)
            .put("presignedUploadUrl", presignedUploadUrl)
            .put("objectStorageAccessKeyId", objectStorageAccessKeyId)
            .put("objectStorageSecretAccessKey", objectStorageSecretAccessKey)
            .put("objectStorageBucket", objectStorageBucket)
            .put("objectStorageEndpoint", objectStorageEndpoint)
            .put("objectStorageAppId", objectStorageAppId)
            .put("objectStorageCustomUrl", objectStorageCustomUrl)
            .put("objectStorageObjectKey", objectStorageObjectKey)
            .put("autoSyncEnabled", autoSyncEnabled)
            .put("autoSyncIntervalMinutes", autoSyncIntervalMinutes)
            .put("autoSyncIntervalValue", autoSyncIntervalValue)
            .put("autoSyncIntervalUnit", autoSyncIntervalUnit.wireName)
            .put("autoSyncOnUnlock", autoSyncOnUnlock)
            .put("conflictStrategy", conflictStrategy.wireName)
            .put("syncMasterKey", syncMasterKey)
            .put("deviceId", deviceId)
            .put("lastSyncRevision", lastSyncRevision)
            .put("lastSyncAt", lastSyncAt?.toString())
            .put("lastSyncStatus", lastSyncStatus)
            .put("lastSyncMessage", lastSyncMessage)
            .put("lastRemoteFingerprint", lastRemoteFingerprint)
            .put("hasLocalChanges", hasLocalChanges)
            .put(
                "logs",
                JSONArray(
                    logs.map { entry ->
                        JSONObject()
                            .put("timestamp", entry.timestamp.toString())
                            .put("message", entry.message)
                            .put("level", entry.level)
                    }
                )
            )

    companion object {
        fun defaults(deviceId: String = generateDeviceId()): SyncSettings =
            SyncSettings(
                providerType = SyncProviderType.NONE,
                webdavUrl = "",
                webdavUsername = "",
                webdavPassword = "",
                webdavPath = "/vault.json",
                presignedDownloadUrl = "",
                presignedUploadUrl = "",
                objectStorageAccessKeyId = "",
                objectStorageSecretAccessKey = "",
                objectStorageBucket = "",
                objectStorageEndpoint = "",
                objectStorageAppId = "",
                objectStorageCustomUrl = "",
                objectStorageObjectKey = "vault.sync.json",
                autoSyncEnabled = false,
                autoSyncIntervalMinutes = 30,
                autoSyncIntervalValue = 30,
                autoSyncIntervalUnit = SyncIntervalUnit.MINUTES,
                autoSyncOnUnlock = true,
                conflictStrategy = SyncSettingsConflictStrategy.REMOTE_WINS,
                syncMasterKey = true,
                deviceId = deviceId,
                lastSyncRevision = 0,
                lastSyncAt = null,
                lastSyncStatus = null,
                lastSyncMessage = null,
                lastRemoteFingerprint = null,
                hasLocalChanges = false,
                logs = emptyList(),
            )

        fun fromJson(json: JSONObject): SyncSettings {
            val defaults = defaults(deviceId = "")
            val intervalUnit = SyncIntervalUnit.fromWireName(json.optionalString("autoSyncIntervalUnit"))
            val intervalValue = if (json.has("autoSyncIntervalValue") && !json.isNull("autoSyncIntervalValue")) {
                json.optInt("autoSyncIntervalValue", defaults.autoSyncIntervalValue)
            } else {
                json.optInt("autoSyncIntervalMinutes", defaults.autoSyncIntervalMinutes)
            }.coerceAtLeast(1)
            val intervalMinutes = if (json.has("autoSyncIntervalMinutes") && !json.isNull("autoSyncIntervalMinutes")) {
                json.optInt("autoSyncIntervalMinutes", defaults.autoSyncIntervalMinutes)
            } else {
                intervalValue.toIntervalMinutes(intervalUnit)
            }.coerceAtLeast(1)
            return SyncSettings(
                providerType = SyncProviderType.fromWireName(json.optionalString("providerType")),
                webdavUrl = json.optString("webdavUrl", defaults.webdavUrl),
                webdavUsername = json.optString("webdavUsername", defaults.webdavUsername),
                webdavPassword = json.optString("webdavPassword", defaults.webdavPassword),
                webdavPath = json.optString("webdavPath", defaults.webdavPath),
                presignedDownloadUrl = json.optString("presignedDownloadUrl", defaults.presignedDownloadUrl),
                presignedUploadUrl = json.optString("presignedUploadUrl", defaults.presignedUploadUrl),
                objectStorageAccessKeyId = json.optString("objectStorageAccessKeyId", defaults.objectStorageAccessKeyId),
                objectStorageSecretAccessKey = json.optString("objectStorageSecretAccessKey", defaults.objectStorageSecretAccessKey),
                objectStorageBucket = json.optString("objectStorageBucket", defaults.objectStorageBucket),
                objectStorageEndpoint = json.optString("objectStorageEndpoint", defaults.objectStorageEndpoint),
                objectStorageAppId = json.optString("objectStorageAppId", defaults.objectStorageAppId),
                objectStorageCustomUrl = json.optString("objectStorageCustomUrl", defaults.objectStorageCustomUrl),
                objectStorageObjectKey = json.optString("objectStorageObjectKey", defaults.objectStorageObjectKey),
                autoSyncEnabled = json.optBoolean("autoSyncEnabled", defaults.autoSyncEnabled),
                autoSyncIntervalMinutes = intervalMinutes,
                autoSyncIntervalValue = intervalValue,
                autoSyncIntervalUnit = intervalUnit,
                autoSyncOnUnlock = json.optBoolean("autoSyncOnUnlock", defaults.autoSyncOnUnlock),
                conflictStrategy = SyncSettingsConflictStrategy.fromWireName(json.optionalString("conflictStrategy")),
                syncMasterKey = json.optBoolean("syncMasterKey", defaults.syncMasterKey),
                deviceId = json.optString("deviceId", defaults.deviceId),
                lastSyncRevision = json.optInt("lastSyncRevision", defaults.lastSyncRevision),
                lastSyncAt = json.optionalString("lastSyncAt")?.let(Instant::parse),
                lastSyncStatus = json.optionalString("lastSyncStatus"),
                lastSyncMessage = json.optionalString("lastSyncMessage"),
                lastRemoteFingerprint = json.optionalString("lastRemoteFingerprint"),
                hasLocalChanges = json.optBoolean("hasLocalChanges", defaults.hasLocalChanges),
                logs = json.optJSONArray("logs")?.toSyncLogs().orEmpty(),
            )
        }

        fun generateDeviceId(): String =
            "${Instant.now().toEpochMilli() * 1000}-${UUID.randomUUID().toString().take(8).lowercase(Locale.US)}"
    }
}

fun Int.toIntervalMinutes(unit: SyncIntervalUnit): Int =
    when (unit) {
        SyncIntervalUnit.SECONDS -> ((this.coerceAtLeast(1) + 59) / 60).coerceAtLeast(1)
        SyncIntervalUnit.MINUTES -> coerceAtLeast(1)
    }

data class SyncSecretBundle(
    val webdavPassword: String = "",
    val presignedDownloadUrl: String = "",
    val presignedUploadUrl: String = "",
    val objectStorageAccessKeyId: String = "",
    val objectStorageSecretAccessKey: String = "",
) {
    val isEmpty: Boolean
        get() = webdavPassword.isEmpty() &&
            presignedDownloadUrl.isEmpty() &&
            presignedUploadUrl.isEmpty() &&
            objectStorageAccessKeyId.isEmpty() &&
            objectStorageSecretAccessKey.isEmpty()

    companion object {
        val EMPTY = SyncSecretBundle()
    }
}

class SyncClientFactory {
    fun makeClient(
        settings: SyncSettings,
        transport: RemoteSyncHttpTransport = UrlConnectionRemoteSyncTransport(),
    ): RemoteSyncClient? =
        when (settings.providerType) {
            SyncProviderType.NONE -> null
            SyncProviderType.WEBDAV,
            SyncProviderType.NAS_WEBDAV -> {
                if (settings.webdavUrl.isBlank() || settings.webdavPath.isBlank()) {
                    null
                } else {
                    WebDavSyncClient(
                        baseUrl = settings.webdavUrl.trim(),
                        remotePath = settings.webdavPath.trim(),
                        username = settings.webdavUsername.trim(),
                        password = settings.webdavPassword.trim(),
                        transport = transport,
                    )
                }
            }
            SyncProviderType.S3_PRESIGNED -> {
                if (settings.presignedUploadUrl.isBlank()) {
                    null
                } else {
                    PresignedUrlSyncClient(
                        downloadUrl = settings.presignedDownloadUrl.trim(),
                        uploadUrl = settings.presignedUploadUrl.trim(),
                        transport = transport,
                    )
                }
            }
            SyncProviderType.TENCENT_COS,
            SyncProviderType.ALIYUN_OSS -> {
                if (
                    settings.objectStorageAccessKeyId.isBlank() ||
                    settings.objectStorageSecretAccessKey.isBlank() ||
                    settings.objectStorageBucket.isBlank() ||
                    (settings.objectStorageEndpoint.isBlank() && settings.objectStorageCustomUrl.isBlank())
                ) {
                    null
                } else {
                    ObjectStorageSyncClient(
                        providerType = settings.providerType,
                        accessKeyId = settings.objectStorageAccessKeyId.trim(),
                        secretAccessKey = settings.objectStorageSecretAccessKey.trim(),
                        bucket = settings.objectStorageBucket.trim(),
                        endpoint = settings.objectStorageEndpoint.trim(),
                        appId = settings.objectStorageAppId.trim(),
                        customUrl = settings.objectStorageCustomUrl.trim(),
                        objectKey = settings.objectStorageObjectKey.trim().ifBlank { "vault.sync.json" },
                        transport = transport,
                    )
                }
            }
        }
}

private fun JSONArray.toSyncLogs(): List<SyncLogEntry> =
    (0 until length()).mapNotNull { index ->
        val item = optJSONObject(index) ?: return@mapNotNull null
        val timestamp = item.optionalString("timestamp")
            ?.let(Instant::parse)
            ?: Instant.EPOCH
        SyncLogEntry(
            timestamp = timestamp,
            message = item.optString("message", ""),
            level = item.optString("level", "info"),
        )
    }

private fun JSONObject.optionalString(name: String): String? =
    if (has(name) && !isNull(name)) {
        optString(name).takeUnless { it.isBlank() }
    } else {
        null
    }

File: apps/ios_native/Sources/PasswordManageriOSCore/Sync/SyncSettings.swift
import Foundation

enum SyncProviderType: String, CaseIterable, Codable, Sendable {
    case none
    case webdav
    case s3Presigned
    case nasWebdav

    var title: String {
        switch self {
        case .none:
            "None"
        case .webdav:
            "WebDAV"
        case .s3Presigned:
            "S3 Presigned URL"
        case .nasWebdav:
            "NAS WebDAV"
        }
    }
}

enum SyncSettingsConflictStrategy: String, CaseIterable, Codable, Sendable {
    case remoteWins
    case localWins
    case keepBoth

    var title: String {
        switch self {
        case .remoteWins:
            "Remote Wins"
        case .localWins:
            "Local Wins"
        case .keepBoth:
            "Keep Both"
        }
    }

    var mergeStrategy: SyncConflictStrategy {
        switch self {
        case .remoteWins:
            .remoteWins
        case .localWins:
            .localWins
        case .keepBoth:
            .keepBoth
        }
    }
}

enum SyncIntervalUnit: String, CaseIterable, Codable, Sendable {
    case seconds
    case minutes

    var title: String {
        switch self {
        case .seconds:
            "Seconds"
        case .minutes:
            "Minutes"
        }
    }
}

struct SyncLogEntry: Codable, Equatable, Sendable {
    var timestamp: Date
    var message: String
    var level: String
}

struct SyncSettings: Codable, Equatable, Sendable {
    var providerType: SyncProviderType
    var webdavUrl: String
    var webdavUsername: String
    var webdavPassword: String
    var webdavPath: String
    var presignedDownloadUrl: String
    var presignedUploadUrl: String
    var autoSyncEnabled: Bool
    var autoSyncIntervalMinutes: Int
    var autoSyncIntervalValue: Int
    var autoSyncIntervalUnit: SyncIntervalUnit
    var autoSyncOnUnlock: Bool
    var conflictStrategy: SyncSettingsConflictStrategy
    var syncMasterKey: Bool
    var deviceId: String
    var lastSyncRevision: Int
    var lastSyncAt: Date?
    var lastSyncStatus: String?
    var lastSyncMessage: String?
    var lastRemoteFingerprint: String?
    var hasLocalChanges: Bool
    var logs: [SyncLogEntry]

    private enum CodingKeys: String, CodingKey {
        case providerType
        case webdavUrl
        case webdavUsername
        case webdavPassword
        case webdavPath
        case presignedDownloadUrl
        case presignedUploadUrl
        case autoSyncEnabled
        case autoSyncIntervalMinutes
        case autoSyncIntervalValue
        case autoSyncIntervalUnit
        case autoSyncOnUnlock
        case conflictStrategy
        case syncMasterKey
        case deviceId
        case lastSyncRevision
        case lastSyncAt
        case lastSyncStatus
        case lastSyncMessage
        case lastRemoteFingerprint
        case hasLocalChanges
        case logs
    }

    static func defaults(deviceId: String = generateDeviceId()) -> SyncSettings {
        SyncSettings(
            providerType: .none,
            webdavUrl: "",
            webdavUsername: "",
            webdavPassword: "",
            webdavPath: "/vault.json",
            presignedDownloadUrl: "",
            presignedUploadUrl: "",
            autoSyncEnabled: false,
            autoSyncIntervalMinutes: 30,
            autoSyncIntervalValue: 30,
            autoSyncIntervalUnit: .minutes,
            autoSyncOnUnlock: true,
            conflictStrategy: .remoteWins,
            syncMasterKey: true,
            deviceId: deviceId,
            lastSyncRevision: 0,
            lastSyncAt: nil,
            lastSyncStatus: nil,
            lastSyncMessage: nil,
            lastRemoteFingerprint: nil,
            hasLocalChanges: false,
            logs: []
        )
    }

    static func generateDeviceId() -> String {
        "\(Int(Date().timeIntervalSince1970 * 1_000_000))-\(UUID().uuidString.prefix(8).lowercased())"
    }

    init(
        providerType: SyncProviderType,
        webdavUrl: String,
        webdavUsername: String,
        webdavPassword: String,
        webdavPath: String,
        presignedDownloadUrl: String,
        presignedUploadUrl: String,
        autoSyncEnabled: Bool,
        autoSyncIntervalMinutes: Int,
        autoSyncIntervalValue: Int,
        autoSyncIntervalUnit: SyncIntervalUnit,
        autoSyncOnUnlock: Bool,
        conflictStrategy: SyncSettingsConflictStrategy,
        syncMasterKey: Bool,
        deviceId: String,
        lastSyncRevision: Int,
        lastSyncAt: Date?,
        lastSyncStatus: String?,
        lastSyncMessage: String?,
        lastRemoteFingerprint: String?,
        hasLocalChanges: Bool,
        logs: [SyncLogEntry]
    ) {
        self.providerType = providerType
        self.webdavUrl = webdavUrl
        self.webdavUsername = webdavUsername
        self.webdavPassword = webdavPassword
        self.webdavPath = webdavPath
        self.presignedDownloadUrl = presignedDownloadUrl
        self.presignedUploadUrl = presignedUploadUrl
        self.autoSyncEnabled = autoSyncEnabled
        self.autoSyncIntervalMinutes = autoSyncIntervalMinutes
        self.autoSyncIntervalValue = autoSyncIntervalValue
        self.autoSyncIntervalUnit = autoSyncIntervalUnit
        self.autoSyncOnUnlock = autoSyncOnUnlock
        self.conflictStrategy = conflictStrategy
        self.syncMasterKey = syncMasterKey
        self.deviceId = deviceId
        self.lastSyncRevision = lastSyncRevision
        self.lastSyncAt = lastSyncAt
        self.lastSyncStatus = lastSyncStatus
        self.lastSyncMessage = lastSyncMessage
        self.lastRemoteFingerprint = lastRemoteFingerprint
        self.hasLocalChanges = hasLocalChanges
        self.logs = logs
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let defaults = Self.defaults(deviceId: "")
        providerType = Self.decodeEnum(
            SyncProviderType.self,
            from: container,
            forKey: .providerType,
            defaultValue: defaults.providerType
        )
        webdavUrl = try container.decodeIfPresent(String.self, forKey: .webdavUrl) ?? defaults.webdavUrl
        webdavUsername = try container.decodeIfPresent(String.self, forKey: .webdavUsername) ?? defaults.webdavUsername
        webdavPassword = try container.decodeIfPresent(String.self, forKey: .webdavPassword) ?? defaults.webdavPassword
        webdavPath = try container.decodeIfPresent(String.self, forKey: .webdavPath) ?? defaults.webdavPath
        presignedDownloadUrl = try container.decodeIfPresent(String.self, forKey: .presignedDownloadUrl) ?? defaults.presignedDownloadUrl
        presignedUploadUrl = try container.decodeIfPresent(String.self, forKey: .presignedUploadUrl) ?? defaults.presignedUploadUrl
        autoSyncEnabled = try container.decodeIfPresent(Bool.self, forKey: .autoSyncEnabled) ?? defaults.autoSyncEnabled
        autoSyncIntervalUnit = Self.decodeEnum(
            SyncIntervalUnit.self,
            from: container,
            forKey: .autoSyncIntervalUnit,
            defaultValue: defaults.autoSyncIntervalUnit
        )
        let decodedIntervalValue = try container.decodeIfPresent(Int.self, forKey: .autoSyncIntervalValue)
            ?? container.decodeIfPresent(Int.self, forKey: .autoSyncIntervalMinutes)
            ?? defaults.autoSyncIntervalValue
        autoSyncIntervalValue = Swift.max(decodedIntervalValue, 1)
        let decodedIntervalMinutes = try container.decodeIfPresent(Int.self, forKey: .autoSyncIntervalMinutes)
            ?? autoSyncIntervalValue.toIntervalMinutes(unit: autoSyncIntervalUnit)
        autoSyncIntervalMinutes = Swift.max(decodedIntervalMinutes, 1)
        autoSyncOnUnlock = try container.decodeIfPresent(Bool.self, forKey: .autoSyncOnUnlock) ?? defaults.autoSyncOnUnlock
        conflictStrategy = Self.decodeEnum(
            SyncSettingsConflictStrategy.self,
            from: container,
            forKey: .conflictStrategy,
            defaultValue: defaults.conflictStrategy
        )
        syncMasterKey = try container.decodeIfPresent(Bool.self, forKey: .syncMasterKey) ?? defaults.syncMasterKey
        deviceId = try container.decodeIfPresent(String.self, forKey: .deviceId) ?? ""
        lastSyncRevision = try container.decodeIfPresent(Int.self, forKey: .lastSyncRevision) ?? defaults.lastSyncRevision
        lastSyncAt = try container.decodeIfPresent(Date.self, forKey: .lastSyncAt)
        lastSyncStatus = try container.decodeIfPresent(String.self, forKey: .lastSyncStatus)
        lastSyncMessage = try container.decodeIfPresent(String.self, forKey: .lastSyncMessage)
        lastRemoteFingerprint = try container.decodeIfPresent(String.self, forKey: .lastRemoteFingerprint)
        hasLocalChanges = try container.decodeIfPresent(Bool.self, forKey: .hasLocalChanges) ?? defaults.hasLocalChanges
        logs = try container.decodeIfPresent([SyncLogEntry].self, forKey: .logs) ?? defaults.logs
    }

    private static func decodeEnum<T: RawRepresentable>(
        _ type: T.Type,
        from container: KeyedDecodingContainer<CodingKeys>,
        forKey key: CodingKeys,
        defaultValue: T
    ) -> T where T.RawValue == String {
        guard let rawValue = try? container.decodeIfPresent(String.self, forKey: key),
              let decoded = T(rawValue: rawValue) else {
            return defaultValue
        }
        return decoded
    }

    var syncSecrets: SyncSecretBundle {
        SyncSecretBundle(
            webdavPassword: webdavPassword,
            presignedDownloadUrl: presignedDownloadUrl,
            presignedUploadUrl: presignedUploadUrl
        )
    }

    func redactedForPlaintextStorage() -> SyncSettings {
        applyingSecrets(.empty)
    }

    func applyingSecrets(_ secrets: SyncSecretBundle) -> SyncSettings {
        var copy = self
        copy.webdavPassword = secrets.webdavPassword
        copy.presignedDownloadUrl = secrets.presignedDownloadUrl
        copy.presignedUploadUrl = secrets.presignedUploadUrl
        return copy
    }
}

extension Int {
    func toIntervalMinutes(unit: SyncIntervalUnit) -> Int {
        switch unit {
        case .seconds:
            Swift.max((Swift.max(self, 1) + 59) / 60, 1)
        case .minutes:
            Swift.max(self, 1)
        }
    }
}

struct SyncSecretBundle: Codable, Equatable, Sendable {
    var webdavPassword: String = ""
    var presignedDownloadUrl: String = ""
    var presignedUploadUrl: String = ""

    static let empty = SyncSecretBundle()

    var isEmpty: Bool {
        webdavPassword.isEmpty &&
            presignedDownloadUrl.isEmpty &&
            presignedUploadUrl.isEmpty
    }
}

struct SyncClientFactory: Sendable {
    func makeClient(
        settings: SyncSettings,
        transport: RemoteSyncHTTPTransport = URLSessionRemoteSyncTransport()
    ) -> RemoteSyncClient? {
        switch settings.providerType {
        case .none:
            return nil
        case .webdav, .nasWebdav:
            guard !settings.webdavUrl.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
                  !settings.webdavPath.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
                return nil
            }
            return WebDavSyncClient(
                baseUrl: settings.webdavUrl.trimmingCharacters(in: .whitespacesAndNewlines),
                remotePath: settings.webdavPath.trimmingCharacters(in: .whitespacesAndNewlines),
                username: settings.webdavUsername.trimmingCharacters(in: .whitespacesAndNewlines),
                password: settings.webdavPassword.trimmingCharacters(in: .whitespacesAndNewlines),
                transport: transport
            )
        case .s3Presigned:
            guard !settings.presignedUploadUrl.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
                return nil
            }
            return PresignedUrlSyncClient(
                downloadUrl: settings.presignedDownloadUrl.trimmingCharacters(in: .whitespacesAndNewlines),
                uploadUrl: settings.presignedUploadUrl.trimmingCharacters(in: .whitespacesAndNewlines),
                transport: transport
            )
        }
    }
}
